File: py/camera.py
#!/usr/bin/env python
# coding: utf-8

import cv2
import threading


class SimpleCamera(object):

    def __init__(self, video_src=0):
        self._cam = None
        self._fps = None
        self._frame = None
        self._frame_width = None
        self._frame_height = None
        self._ret = False
        self._video_src = video_src
        self._stop_event = threading.Event()

    def _init_camera(self):
        self._cam = cv2.VideoCapture(self._video_src)
        self._fps = self._cam.get(cv2.CAP_PROP_FPS)
        self._ret, self._frame = self._cam.read()
        if not self._ret:
            raise Exception("No camera feed")
        self._frame_height, self._frame_width, c = self._frame.shape
        return self._ret

    def start_camera(self):
        """
        Start the running of the camera, without this we can't capture frames
        Camera runs on a separate thread so we can reach a higher FPS
        """
        self._init_camera()
        self._stop_event.clear()
        threading.Thread(target=self._update_camera, args=()).start()

    def read(self):
        """
        With this you can grab the last frame from the camera
        :return (boolean, np.array): return value and frame
        """
        if self.is_running:
            return self._ret, self._frame
        else:
            import warnings
            warnings.warn("Camera is not started, you should start it with start_camera()")
            return False, None

    def _update_camera(self):
        """
        Grabs the frames from the camera
        """
        while self.is_running:
            ret, frm = self._cam.read()
            self._ret, self._frame = (True, frm) if ret else (False, None)

    def get_frame_width_and_height(self):
        """
        Returns the width and height of the grabbed images
        :return (int int): width and height
        """
        return self._frame_width, self._frame_height

    def get_fps(self):
        return self._fps

    def release_camera(self):
        """
        Stop the camera
        """
        self._stop_event.set()
        # self._cam.release()

    @property
    def is_running(self):
        return not self._stop_event.is_set()

File: cpp/camera.cpp
#include <iostream>
#include <thread>
#include <chrono>
#include <stdexcept>

#include "spdlog/spdlog.h"

#include "camera.hpp"


namespace spd = spdlog;

void SimpleCamera::start_camera() {
    auto logger = spd::get("toy");
    logger->info("start camera");

    bool ret = init_camera();
    if (!ret) {
        logger->error("Failed to init camera");
        throw std::runtime_error("Failed to init camera");
    }
    logger->debug("ready to start worker");

    std::hash<std::thread::id> hasher;
    logger->debug("_worker:{} before start", hasher(_worker.get_id()));
    _worker = std::thread(&SimpleCamera::update_camera, this);
    logger->info("_worker:{} started", hasher(_worker.get_id()));
    _worker.detach();
} // SimpleCamera::start_camera


bool SimpleCamera::init_camera() {
    auto logger = spd::get("toy");

    if (!_cam.isOpened()) {
        logger->error("Failed to open the video device or video file!");
        return false;
    }
    logger->debug("_cam opened");

    _fps = _cam.get(cv::CAP_PROP_FPS);
    _frame_size.width = _cam.get(cv::CAP_PROP_FRAME_WIDTH);
    _frame_size.height = _cam.get(cv::CAP_PROP_FRAME_HEIGHT);
    // limit the exposure time
    // _cam.set(cv::CAP_PROP_EXPOSURE, 40);
    logger->debug("width:{}, height:{}", _frame_size.width, _frame_size.height);

    {
        std::lock_guard<std::mutex> lg(v_mutex);
        _cam >> _frame;
    }
    logger->debug("====> SimpleCamera read to _frame");

    if (_frame.empty()) {
        logger->error("_frame is empty!!!");
        _frm_ret = false;
    } else {
        _is_running = true;
        _frm_ret = true;
    }

    return _frm_ret;
} // SimpleCamera::init_camera


void SimpleCamera::update_camera() {
    auto logger = spd::get("toy");

    while (_is_running) {
        cv::Mat img_buf;
        _cam >> img_buf;
        if (!img_buf.empty()) {
            std::lock_guard<std::mutex> lg(v_mutex);
            img_buf.copyTo(_frame);
        } else {
            // NOTE:
            img_buf.copyTo(_frame);
            _frm_ret = false;
            logger->warn("img_buf empty!!!");
        }

        std::this_thread::yield();
        // std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
} // SimpleCamera::update_camera
